pages/admin/students.js

```diff
@@ -0,0 +1,236 @@
+import Button from "react-bootstrap/Button";
+import Form from "react-bootstrap/Form";
+import FormControl from "react-bootstrap/FormControl";
+import BootstrapTable from "react-bootstrap-table-next";
+import Layout from "../../components/Layout";
+import { createRequiredAuth } from "../../utils/ssr";
+import { serializeDocument } from "../../utils/mongodb";
+import { useToasts } from "../../components/Toasts";
+import { getStudents } from "../api/students";
+import ReactFileReader from "react-file-reader";
+import { reformatEmail } from "../../utils/email";
+
+export const getServerSideProps = async ({ req, res }) => {
+  const ssr = await createRequiredAuth({ roles: ["admin"] })({ req, res });
+
+  ssr.props.initialData = (await getStudents()).map(serializeDocument);
+
+  return ssr;
+};
+
+function getColumnsWithActions(actionsFn) {
+  return [
+    {
+      dataField: "fname",
+      text: "First Name",
+    },
+    {
+      dataField: "lname",
+      text: "Last Name",
+    },
+    {
+      dataField: "permNum",
+      text: "Perm",
+    },
+    {
+      dataField: "email",
+      text: "Email",
+    },
+    {
+      dataField: "section",
+      text: "Section",
+    },
+    {
+      dataField: "df1",
+      isDummyField: true,
+      text: "Actions",
+      formatter: actionsFn,
+    },
+  ];
+}
+
+export default function ManageStudentsPage(props) {
+  const { user, initialData } = props;
+  const { showToast } = useToasts();
+  const { data, mutate } = useSWR("/api/students", { initialData });
+  const [newStudentEmail, setNewStudentEmail] = useState("");
+  const [newStudentSection, setNewStudentSection] = useState("");
+  const [newStudentFname, setNewStudentFname] = useState("");
+  const [newStudentLname, setNewStudentLname] = useState("");
+  const [newStudentPermNum, setNewStudentPermNum] = useState("");
+  const addStudent = useCallback(
+    async (e) => {
+      // override default form submission behavior
+      e.preventDefault();
+      e.stopPropagation();
+
+      setNewStudentEmail("");
+      setNewStudentSection("");
+      setNewStudentFname("");
+      setNewStudentLname("");
+      setNewStudentPermNum("");
+      showToast(`Added student ${reformatEmail(newStudentEmail)}`);
+      await mutate(
+        [
+          ...data,
+          {
+            email: reformatEmail(newStudentEmail),
+            section: newStudentSection,
+            fname: newStudentFname,
+            lname: newStudentLname,
+            permNum: newStudentPermNum,
+          },
+        ],
+        false
+      );
+      await fetch("/api/students", {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          email: reformatEmail(newStudentEmail),
+          section: newStudentSection,
+          fname: newStudentFname,
+          lname: newStudentLname,
+          permNum: newStudentPermNum,
+        }),
+      });
+      await mutate();
```

Comment:
Is there any particular reason we have this trailing mutate here? It doesn't appear to be updating any cached state.

Reply:
@Ulramok that's a question for you